### PR TITLE
[pytorch][nnc] mobile nnc backend skeleton

### PR DESCRIPTION
**Pull request by ljk53**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#56852 [pytorch][nnc] mobile nnc backend skeleton**
* #56851 [pytorch][nnc] protocol classes to persist the context for compiled functions

This is part of the changes to enable NNC AOT compilation for mobile.
It introduced a custom backend for NNC, which uses the components defined in the stacked PRs to load and execute a NNC-compiled model.

Differential Revision: [D27897153](https://our.internmc.facebook.com/intern/diff/D27897153/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27897153/)!

Differential Revision: [D27897153](https://our.internmc.facebook.com/intern/diff/D27897153)

### CODE REVIEW

File: torch/csrc/jit/mobile/nnc/backend.cpp [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    auto function_name = c10::QualifiedName(method_name);
+    return cu_->run(function_name, inputs);
+  }
+
+ private:
+  std::shared_ptr<CompilationUnit> cu_;
+};
+
+namespace {
+static const auto cls = torch::jit::backend<NNCBackend>("nnc");
```
Comment (kimishpatel) [inline torch/csrc/jit/mobile/nnc/backend.cpp:55]: What about the corresponding preprocess function?

**Review by kimishpatel — COMMENTED** [review]
Dont see the preprocess function?

File: torch/csrc/jit/mobile/nnc/backend.cpp (hunk 1 above)
Comment (ljk53) [reply torch/csrc/jit/mobile/nnc/backend.cpp:55]: I think the new interface allows us to register the backend without preprocess method. And I think it's correct - the runtime binary itself doesn't need to have code to do AOT compilation.
The AOT part is in the stacked Phabricator diff: D27897136

Comment (kimishpatel) [reply torch/csrc/jit/mobile/nnc/backend.cpp:55]: > The AOT part is in the stacked Phabricator diff: D27897136

Thats what I was looking for. Thanks!

**Review by kimishpatel — APPROVED** [review]
Looks good to me.

**Review by iseeyuan — APPROVED** [review]
LGTM

**Review by raziel — APPROVED** [review]

**Pull request closed without merging**